### ISSUE TITLE
Have to force quit emacs when typing <

### ISSUE THREAD

**Issue by dfrkp** [felipeochoa/rjsx-mode#50]
Hi,

I'm on macOS, using GNU Emacs 25.2.1 (x86_64-apple-darwin13.4.0, NS appkit-1265.21 Version 10.9.5 (Build 13F1911)) of 2017-04-21 and loading rjsx-mode like this:
`(req-package rjsx-mode
  :config
  (add-to-list 'auto-mode-alist '("\\.jsx?$" . rjsx-mode)))`
When I enter a render function and type "<" emacs freezes and I have to force quit.
For now I'm fixing it like this:
`(req-package rjsx-mode
  :config
  (add-to-list 'auto-mode-alist '("\\.jsx?$" . rjsx-mode))
  (with-eval-after-load 'rjsx-mode
    (define-key rjsx-mode-map "<" nil)
    (define-key rjsx-mode-map (kbd "C-d") nil)))`
However, I'd like to use the "<" feature. So I'm looking forward to a fix ;-) Let me know if I can provide additional debugging information.

**Comment by felipeochoa** [comment]
Thanks for the report. What version of `rjsx-mode` are you using? If you open a fresh emacs with "-Q" does the problem persist?

**Comment by dfrkp** [comment]
@felipeochoa , I'm using the melpa current - I believe - >> 20170928.132
When I use -Q rjsx-mode isn't loaded so it works fine - sorry, not sure what I should do there

**Comment by felipeochoa** [comment]
After running `emacs -Q` evaluate the following to load rjsx and js2.

```elisp
(dolist (m '("js2" "rjsx"))
  (load-file
   (concat
    "~/.emacs.d/elpa/"
    (car (directory-files "~/.emacs.d/elpa/" nil (concat m "-mode-.*")))
    "/" m "-mode.el")))
```

(This assumes you have `js2-mode-XXX/js2-mode.el` and `rjsx-mode-XXX/rjsx-mode.el` under `~/.emacs.d/elpa/`)

You can then visit a JSX file and `rjsx` should autoload.

**Comment by dfrkp** [comment]
Sorry, elisp isn't really my favorite ;-) so I may have some basic issue of understanding. Here's the result from executing the snippet.
```
Debugger entered--Lisp error: (file-error "Cannot open load file" "cl-lib")
  require(cl-lib)
  eval-buffer(#<buffer  *load*> nil "/Users/user/.emacs.d/elpa/js2-mode-20170815.1415/js2-mode.el" nil t)  ; Reading at buffer position 3816
  load-with-code-conversion("/Users/user/.emacs.d/elpa/js2-mode-20170815.1415/js2-mode.el" "/Users/user/.emacs.d/elpa/js2-mode-20170815.1415/js2-mode.el" nil nil)
  load("/Users/user/.emacs.d/elpa/js2-mode-20170815.1415/js2-mode.el" nil nil t)
  load-file("~/.emacs.d/elpa/js2-mode-20170815.1415/js2-mode.el")
  (while --dolist-tail-- (setq m (car --dolist-tail--)) (load-file (concat "~/.emacs.d/elpa/" ... "/" m "-mode.el")) (setq --dolist-tail-- (cdr --dolist-tail--)))
  (let ((--dolist-tail-- ...) m) (while --dolist-tail-- (setq m ...) (load-file ...) (setq --dolist-tail-- ...)))
  (dolist (m (quote ...)) (load-file (concat "~/.emacs.d/elpa/" ... "/" m "-mode.el")))
  eval((dolist (m (quote ...)) (load-file (concat "~/.emacs.d/elpa/" ... "/" m "-mode.el"))))
  eval-last-sexp-1(nil)
  eval-last-sexp(nil)
  call-interactively(eval-last-sexp)
```

I do find cl-lib here
```
/Applications/Emacs.app/Contents/Resources/lisp/emacs-lisp/cl-lib.el.gz
/Applications/Emacs.app/Contents/Resources/lisp/emacs-lisp/cl-lib.elc
/Users/user/.emacs.d/el-get/el-get/recipes/cl-lib.rcp
```

**Comment by felipeochoa** [comment]
Hmm that's strange. `cl-lib` should be part of emacs if you're on 25.2. Can you run another emacs -Q and then check the value of `load-path`? (either evaluate it as before, or do C-h v).

**Comment by dfrkp** [comment]
Ok - I'm sorry for that! Typing emacs on the terminal was using the v22 emacs that is bundled with macOS....
Now running the correct emacs leads to the evaluation working - resulting with nil. Now when I open the jsx file and activate rjsx-mode I can type the "<" without problems, although the "C-d" doesn't do anything. Only thing I noticed was "Code has no side effects" in *Messages*. Not sure if that helps at all.

**Comment by felipeochoa** [comment]
"Code has no side effects" is a warning from `js2-mode` about your js code. In your emacs -Q with rjsx can you check what  "<" and "C-d" are bound ot? (You can use `C-h k <` and `C-h k C-d`)

**Comment by dfrkp** [comment]
C-d runs the command rjsx-delete-creates-full-tag (found in
rjsx-mode-map), which is an interactive Lisp closure in
‘~/.emacs.d/elpa/rjsx-mode-20170928.132/rjsx-mode.el’.

It is bound to C-d.

(rjsx-delete-creates-full-tag N &optional KILLFLAG)

< runs the command rjsx-electric-lt (found in rjsx-mode-map), which is
an interactive Lisp closure in
‘~/.emacs.d/elpa/rjsx-mode-20170928.132/rjsx-mode.el’.

It is bound to <.

(rjsx-electric-lt N)

**Comment by felipeochoa** [comment]
Hmm. Ok. Can you post a sample file that's giving you problems along with the characters you are typing and what the result is?

**Comment by dfrkp** [comment]
Alright I figured something out - when I tried the mode with emacs -Q I didn't pay attention and added a "<Child>" element as the second element in the render function. So it turns out everything works as expected when I add the <Child> above the existing <div or inside it, but not below. Which is fine, since it's anyways wrong syntax when adding a second element.

So do you think there is a chance to understand why it works with emacs -Q but not in my regular setup? Btw. here's the js file I'm using right now: 

```
import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';

class App extends Component {
    render() {
	return (
	    <div className="App">
              <header className="App-header">
		<img src={logo} className="App-logo" alt="logo" />
		<h1 className="App-title">Welcome to React</h1>
              </header>
              <p className="App-intro">
		To get started, edit <code>src/App.js</code> and save to reload.
              </p>
	    </div>
	);
    }
}

export default App;
```

**Comment by felipeochoa** [comment]
Ok, that's good to hear. What you say makes sense based on the way the js2/rjsx parser works. After `return (` it's expecting an expression, so the first JSX element is parsed fine. However, after a JSX element it expects an operator or `)`, so a second JSX element is not able to be parsed.

Since it's not an `rjsx` bug (or at least not without another package loaded), it's harder for me to help you figure it out. Check out `post-command-hook` and `timer-idle-list` (`M-x eval-expression` each of those) to see what else might be running in your regular session. Try disabling all your minor modes (`C-h m` in the jsx buffer) and adding them back one at a time to see when it breaks. Also compare the two variables and the minor modes between the emacs -Q and the regular session.

I'm going to close this issue, but feel free to keep posting here and I'll try to help out.

**Comment by dfrkp** [comment]
Hi @felipeochoa , I was finally able to find the interferring package, it's projectile. When I disable projectile-mode the force quit issue goes away. I will reach out to the projectile team, but figured you may be interested in this as well. (fyi I'm using projectile 20171009.848)

**Comment by dfrkp** [comment]
after being on the wrong track for a while I have finally identified the interfering package to be fill-column-indicator.
Here's how you can reproduce it in a fresh emacs:
```
(load-file "~/.emacs.d/elpa/fill-column-indicator-20170905.1008/fill-column-indicator.el")
(add-hook 'prog-mode-hook 'fci-mode)
(load-file "~/.emacs.d/elpa/js2-mode-20171014.1229/js2-mode.el")
(load-file "~/.emacs.d/elpa/rjsx-mode-20171010.725/rjsx-mode.el")
```
Question is, which package should be doing something differently?

**Comment by felipeochoa** [comment]
I admire your dedication -- I was able to reproduce the issue! I very nearly gave up on this one since it was very challenging to debug, but I finally figured it out. I've pushed the fix, so it should be on Melpa in a few hours.

Here's the postmortem in case you're curious (or another poor soul has to debug an infinite loop inside an `after-change-function`):

* `rjsx-electric-lt` used `inhibit-changing-match-data` a little sloppily and executed `(insert "...")` while inhibiting the change
* `fci` uses an `after-change-function` internally (in `fci-put-overlays-region`) which does:

  ```elisp
  (while (search-forward "\n" end t)
    ;; do stuff...
    (goto-char (match-end 0)))
  ```
* Since `(match-end 0)` never changed, point never advanced and the while loop ran forever
